src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -1395,40 +1397,66 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
         ; basic =
             Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
               (ForeignFieldMul
-                 { (* Current row *) left_input0
+                 { (* left input *)
+                   left_input0
                  ; left_input1
                  ; left_input2
-                 ; right_input0
+                 ; (* right input *) right_input0
                  ; right_input1
                  ; right_input2
-                 ; carry1_lo
-                 ; carry1_hi
-                 ; carry0
-                 ; quotient0
+                 ; (* remainder *) remainder01
+                 ; remainder2
+                 ; (* quotient *) quotient0
                  ; quotient1
                  ; quotient2
-                 ; quotient_bound_carry
-                 ; product1_hi_1
-                 ; (* Next row *) remainder0
-                 ; remainder1
-                 ; remainder2
-                 ; quotient_bound01
-                 ; quotient_bound2
-                 ; product1_lo
+                 ; quotient_hi_bound
+                 ; (* products *) product1_lo
                  ; product1_hi_0
-                 ; (* Coefficients *) foreign_field_modulus0
-                 ; foreign_field_modulus1
-                 ; foreign_field_modulus2
+                 ; product1_hi_1
+                 ; (* carries *) carry0
+                 ; carry1_0
+                 ; carry1_12
+                 ; carry1_24
+                 ; carry1_36
+                 ; carry1_48
+                 ; carry1_60
+                 ; carry1_72
+                 ; carry1_84
+                 ; carry1_86
+                 ; carry1_88
+                 ; carry1_90
+                 ; (* Coefficients *) foreign_field_modulus2
                  ; neg_foreign_field_modulus0
                  ; neg_foreign_field_modulus1
                  ; neg_foreign_field_modulus2
                  } )
         } ) ;
+
+  (*
+   * Add external checks (and related)
+   *)
+  External_checks.append_multi_range_check external_checks
+    (quotient0, quotient1, quotient2) ;
+
+  External_checks.append_multi_range_check external_checks
+    (quotient_hi_bound, product1_lo, product1_hi_0) ;
+
+  (* Instead of appending external check for compact MRC for remainder,
+   * this is added directly, so that the standard limbs
+   * (remainder0, remainder1, remainder2) are copyable in witness cells *)
+  let remainder0, remainder1 =
+    Range_check.compact_multi (module Circuit) remainder01 remainder2
+  in
+
+  External_checks.append_bound_check external_checks ~do_multi_range_check:false
```

Comment:
I think that's more complicated the way you're suggestion because the user would need to remember which ones require the extra multi-range-check and which ones do not at that later time and different place in the code where the context is not obvious.

However, by doing it this way, the user specifies what type of bound check is require in the place where it is required and the context is clear.

Furthermore, the External_checks has the complete information about what precise checks are required and therefore in future this allows full flexibility over how those checks are done because it has the complete information and doesn't require the user to manually remember what to do.

Reply:
All a user will ever have to do in order to constrain the external checks is call `constrain_external_checks`.  The user never needs to do all the complicated work of manually constraining all the external checks.  There is no point to require the user to re-write all that code each time, since they must always be constrained in the same way.  So the user just simply calls `constrain_external_checks`.  This achieves the goal of having a safe and easy to use API.

Reply:
agreed with the `constrain_external_checks` API